dashboard/app/models/circuit_playground_discount_application.rb

```diff
@@ -0,0 +1,73 @@
+#
+# Table name: circuit_playground_discount_applications
+#
+#  id                                  :integer          not null, primary key
+#  user_id                             :integer          not null
+#  unit_6_intention                    :integer
+#  has_confirmed_school                :boolean          default(FALSE), not null
+#  full_discount                       :boolean
+#  admin_set_status                    :boolean          default(FALSE), not null
+#  signature                           :string(255)
+#  signed_at                           :datetime
+#  circuit_playground_discount_code_id :integer
+#  created_at                          :datetime         not null
+#  updated_at                          :datetime         not null
+#
+# Indexes
+#
+#  index_circuit_playground_applications_on_code_id           (circuit_playground_discount_code_id)
+#  index_circuit_playground_discount_applications_on_user_id  (user_id)
+#
+
+class CircuitPlaygroundDiscountApplication < ApplicationRecord
+  belongs_to :user
+  belongs_to :circuit_playground_discount_code
+
+  # @return {boolean} true if (1) Attended CSD TeacherCon '17 (2) are a CSD facilitator
+  def self.user_pd_eligible?(user)
+    csd_cohorts = %w(CSD-TeacherConPhiladelphia CSD-TeacherConPhoenix CSD-TeacherConHouston)
+
+    return true if user.cohorts.any? {|cohort| csd_cohorts.include?(cohort.name)}
+    return true if user.courses_as_facilitator.any? {|course_facilitator| course_facilitator.course == Pd::Workshop::COURSE_CSD}
+    return false
+  end
+
+  # Looks to see if any of the users associated with this studio_person_id are eligibile
+  # for our circuit playground discount
+  def self.studio_person_pd_eligible?(user)
+    User.where(studio_person_id: user.studio_person_id).any? {|associated_user| user_pd_eligible?(associated_user)}
+  end
+
+  # @return {boolean} true if we have at least one section that meets our eligibility
+  #   requirements for student progress
+  def self.student_progress_eligible?(user)
+    user.sections.any?(&:has_sufficient_discount_code_progress?)
+  end
+
+  def self.application_status(user)
+    application = CircuitPlaygroundDiscountApplication.find_by_user_id(user.id)
+
+    status = {
+      # This will be a number from 1-5 (representing which radio button) was selected,
+      # or nil if no selection yet
+      unit_6_intention: application.try(:unit_6_intention),
+      has_confirmed_school: application.try(:has_confirmed_school) || false,
+      # true/false once has_submitted_school is true
+      # false implies partial discount
+      gets_full_discount: application.try(:full_discount),
+      discount_code: application.try(:circuit_playground_discount_code),
+    }
+
+    if application
+      # We won't let you create an application without meeting our eligibility requirements
+      # so no need to check them again if we find an existing application
+      status.merge({is_pd_eligible: true, is_progress_eligible: true})
```

Comment:
Avoid doing our eligibility checks if we've started an application (which shouldn't be possible unless you met the eligibility requirements at some point). This might be a premature optimization.

Reply:
Hard to tell, since it looks like it's not currently possible to create an application at all.  At what point in the process are you expecting an application record to be created?  Right now it's looking like it would happen in one of two places:

1. An admin creates an eligibility override for pd/progress.
2. The user is eligible for pd and progress and submits unit_6_intention.

Reply:
That is my intention.